### PR TITLE
Fix compilation error on gcc 8.x

### PR DESCRIPTION
**Pull request by glipner**
Compiling Velox on GCC 8.x (e.g. Debian 10 used in Google's Dataproc) results in an Internal Compiler Error:
```
/home/user/gluten/ep/build-velox/build/velox_ep/./velox/expression/SimpleFunctionAdapter.h:591:23: internal compiler error: trying to capture ‘this’ in instantiation of generic lambda
               notNull = doApplyNullFree<0>(row, out, readers...);
               ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
This PR modifies the relevant file to avoid this error.

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@glipner There is a CI failure. Run make format-fix to fix and update the PR. 

@laithsakka Laith, would you help review this change?

**Review by laithsakka — APPROVED** [review]

**Comment by mbasmanova** [comment]
@glipner Would you sign the CLA so we can merge the PR?

**Comment by laithsakka** [comment]
@glipner 
Hey to import your PR and land it you need to sign Contributor License Agreement. check the info above at the beginning of the PR

**Comment by glipner** [comment]
CLA signed

**Comment by mbasmanova** [comment]
@glipner I'm getting an error: "This diff is either too old, or sits on an old base revision, making it unsafe to land - try rebasing first." Would you rebase this PR?

**Comment by glipner** [comment]
@mbasmanova I merged `main` from upstream. Is this good enough, or would you rather I rebased?

**Comment by glipner** [comment]
@mbasmanova rebased

**Comment by mbasmanova** [comment]
@glipner I'm getting an error building internally

```
[2023-01-19T07:24:37.144-08:00] buck-out/v2/gen/fbcode/514c7233106db7a8/velox/expression/__velox_expression__/buck-headers/velox/expression/SimpleFunctionAdapter.h:584:19: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
[CONTEXT] [2023-01-19T07:24:37.144-08:00]               [&, this](auto row) INLINE_LAMBDA {
[CONTEXT] [2023-01-19T07:24:37.144-08:00]                   ^
```

**Comment by glipner** [comment]
@mbasmanova I updated my changes to limit them to GCC 8 only

File: velox/expression/SimpleFunctionAdapter.h [hunk 1]
```diff
@@ -511,6 +511,12 @@ class SimpleFunctionAdapter : public VectorFunction {
     }
   }
 
+#if __GNUC__ == 8
+#define FIX_GCC8_BUG , this
```
Comment (mbasmanova) [inline velox/expression/SimpleFunctionAdapter.h:515]: This is a bit ugly. @laithsakka Is there a better way? Do we have to support gcc 8.x?

Comment (glipner) [reply velox/expression/SimpleFunctionAdapter.h:515]: I know 😕

GCC 8 is needed to support Debian 10, which is what Dataproc (Google's EMR) uses.
Using a later GCC version is nontrivial, as there is a versioning issue with `libstc++`.

Another way we can do it is check for the compiler version in the Makefile, and remove `-Wunused-lambda-capture` when it's supported by the compiler (Clang)

Comment (arunthirupathi) [reply velox/expression/SimpleFunctionAdapter.h:515]: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100295

**Pull request closed without merging**